File: blue-core/src/test/java/blue/udo/OpcodeListTest.java
/*
 * blue - object composition environment for csound
 * Copyright (c) 2000-2007 Steven Yi (stevenyi@gmail.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by  the Free Software Foundation; either version 2 of the License or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.LIB.  If not, write to
 * the Free Software Foundation Inc., 59 Temple Place - Suite 330,
 * Boston, MA  02111-1307 USA
 */
package blue.udo;

import electric.xml.Element;
import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

class OpcodeListTest {

    @Test
    void testGetNameOfExistingCopy() {
        OpcodeList list = new OpcodeList();

        UserDefinedOpcode udo1 = new UserDefinedOpcode();
        udo1.setOpcodeName("test1");
        udo1.inTypes = "i";
        udo1.outTypes = "i";
        udo1.codeBody = "code";

        list.addOpcode(udo1);

        UserDefinedOpcode udo2 = new UserDefinedOpcode();
        udo2.setOpcodeName("test2");
        udo2.inTypes = "i";
        udo2.outTypes = "i";
        udo2.codeBody = "code";

        assertEquals("test1", list.getNameOfEquivalentCopy(udo2));

        assertNull(list.getNameOfEquivalentCopy(null));

        udo2.codeBody = "code2";

        assertNull(list.getNameOfEquivalentCopy(udo2));

        udo2.codeBody = "code";
        udo2.inTypes = "k";

        assertNull(list.getNameOfEquivalentCopy(udo2));

        udo2.inTypes = "i";
        udo2.outTypes = "k";

        assertNull(list.getNameOfEquivalentCopy(udo2));
    }

    @Test
    void testGetNameOfExistingModernCopy() {
        OpcodeList list = new OpcodeList();

        UserDefinedOpcode udo1 = new UserDefinedOpcode();
        udo1.style = UDOStyle.MODERN;
        udo1.setOpcodeName("testModern1");
        udo1.inputArguments = "aSig, kDrive";
        udo1.outTypes = "a";
        udo1.codeBody = "aOut = tanh(aSig * kDrive)\nxout aOut";

        list.addOpcode(udo1);

        UserDefinedOpcode udo2 = new UserDefinedOpcode();
        udo2.style = UDOStyle.MODERN;
        udo2.setOpcodeName("testModern2");
        udo2.inputArguments = "aSig, kDrive";
        udo2.outTypes = "a";
        udo2.codeBody = "aOut = tanh(aSig * kDrive)\nxout aOut";

        assertEquals("testModern1", list.getNameOfEquivalentCopy(udo2));

        udo2.inputArguments = "aSig, kOther";
        assertNull(list.getNameOfEquivalentCopy(udo2));
    }

    @Test
    void testLegacyXmlDefaultsToClassicStyle() {
        Element xml = new Element("udo");
        xml.addElement("opcodeName").setText("legacyOpcode");
        xml.addElement("outTypes").setText("a");
        xml.addElement("inTypes").setText("ak");
        xml.addElement("codeBody").setText("aOut = aSig\nxout aOut");
        xml.addElement("comments").setText("legacy");

        UserDefinedOpcode udo = UserDefinedOpcode.loadFromXML(xml);

        assertEquals(UDOStyle.CLASSIC, udo.style);
        assertEquals("legacyOpcode", udo.getOpcodeName());
        assertEquals("ak", udo.inTypes);
        assertEquals("a", udo.outTypes);
    }

}

File: packages/blue-data/tests/udo/opcode-list.test.ts
import { describe, it, expect } from 'vitest';
import { OpcodeList } from '../../src/opcodes/opcode-list';
import { OpcodeDefinition } from '../../src/opcodes/opcode-definition';
import { UDOStyle } from '../../src/opcodes/udo-style';
import { Element } from '../../src/serialization/xml-reader';

describe('OpcodeList', () => {
  it('getNameOfEquivalentCopy finds classic equivalent', () => {
    const list = new OpcodeList();

    const udo1 = new OpcodeDefinition();
    udo1.setName('test1');
    udo1.setStyle(UDOStyle.CLASSIC);
    udo1.setOutTypes('i');
    udo1.setInTypes('i');
    udo1.setCode('code');
    list.addOpcode(udo1);

    // Same code/types, different name
    const udo2 = new OpcodeDefinition();
    udo2.setName('test2');
    udo2.setStyle(UDOStyle.CLASSIC);
    udo2.setOutTypes('i');
    udo2.setInTypes('i');
    udo2.setCode('code');

    expect(list.getNameOfEquivalentCopy(udo2)).toBe('test1');
    expect(list.getNameOfEquivalentCopy(null as unknown as OpcodeDefinition)).toBeNull();

    // Different code
    udo2.setCode('code2');
    expect(list.getNameOfEquivalentCopy(udo2)).toBeNull();

    // Different inTypes
    udo2.setCode('code');
    udo2.setInTypes('k');
    expect(list.getNameOfEquivalentCopy(udo2)).toBeNull();

    // Different outTypes
    udo2.setInTypes('i');
    udo2.setOutTypes('k');
    expect(list.getNameOfEquivalentCopy(udo2)).toBeNull();
  });

  it('getNameOfEquivalentCopy finds modern equivalent', () => {
    const list = new OpcodeList();

    const udo1 = new OpcodeDefinition();
    udo1.setName('testModern1');
    udo1.setStyle(UDOStyle.MODERN);
    udo1.setInputArguments('aSig, kDrive');
    udo1.setOutTypes('a');
    udo1.setCode('aOut = tanh(aSig * kDrive)\nxout aOut');
    list.addOpcode(udo1);

    // Same inputArguments/code/outTypes, different name
    const udo2 = new OpcodeDefinition();
    udo2.setName('testModern2');
    udo2.setStyle(UDOStyle.MODERN);
    udo2.setInputArguments('aSig, kDrive');
    udo2.setOutTypes('a');
    udo2.setCode('aOut = tanh(aSig * kDrive)\nxout aOut');

    expect(list.getNameOfEquivalentCopy(udo2)).toBe('testModern1');

    // Different inputArguments
    udo2.setInputArguments('aSig, kOther');
    expect(list.getNameOfEquivalentCopy(udo2)).toBeNull();
  });

  it('legacy XML defaults to classic style', () => {
    const root = new Element('opcodeList');
    const udoElem = root.addElement('udo');
    udoElem.addElement('opcodeName').setText('legacyOpcode');
    udoElem.addElement('outTypes').setText('a');
    udoElem.addElement('inTypes').setText('ak');
    udoElem.addElement('codeBody').setText('xout 1');
    udoElem.addElement('comments').setText('');

    const list = OpcodeList.loadFromXML(root);
    expect(list.size()).toBe(1);
    const udo = list.getOpcode(0)!;
    expect(udo.getStyle()).toBe(UDOStyle.CLASSIC);
    expect(udo.getName()).toBe('legacyOpcode');
    expect(udo.getInTypes()).toBe('ak');
    expect(udo.getOutTypes()).toBe('a');
  });
});
